packages/eslint-plugin/docs/rules/typedef.md

```diff
@@ -0,0 +1,364 @@
+  return size * 2;
+};
+
+const divider = {
+  dividesSize(size: number) {
+    return size / 2;
+  },
+  doesNothing: function() {
+    console.log('...');
+  },
+};
+
+class Logger {
+  log(text: string) {
+    console.log('>', text);
+    return true;
+  }
+}
+```
+
+Examples of **correct** code with `{ "callSignature": true }`:
+
+```ts
+function logsSize(size: number): void {
+  console.log(size);
+}
+
+const doublesSize = function(size: number): number {
+  return size * 2;
+};
+
+const divider = {
+  dividesSize(size: number): number {
+    return size / 2;
+  },
+  doesNothing: function(): void {
+    console.log('...');
+  },
+};
+
+class Logger {
+  log(text: string): boolean {
+    console.log('>', text);
+    return true;
+  }
+}
+```
+
+### memberVariableDeclaration
+
+Whether to enforce type annotations on member variables of classes.
+
+Examples of **incorrect** code with `{ "callSignature": true }`:
+
+```ts
+class ContainsText {
+  delayedText;
+  immediateTextImplicit = 'text';
+}
+```
+
+Examples of **correct** code with `{ "callSignature": true }`:
+
+```ts
+class ContainsText {
+  delayedText: string;
+  immediateTextImplicit: string = 'text';
+}
+```
+
+### objectDestructuring
+
+Whether to enforce type annotations on variables declared using object destructuring.
+
+Examples of **incorrect** code with `{ "objectDestructuring": true }`:
+
+```ts
+const { length } = 'text';
+const [b, c] = Math.random() ? [1, 2] : [3, 4];
+```
+
+Examples of **correct** code with `{ "objectDestructuring": true }`:
+
+```ts
+const { length }: { length: number } = 'text';
+const [b, c]: [number, number] = Math.random() ? [1, 2] : [3, 4];
+```
+
+### parameter
+
+Whether to enforce type annotations for parameters of functions and methods.
+
+Examples of **incorrect** code with `{ "parameter": true }`:
+
+```ts
+function logsSize(size): void {
+  console.log(size);
+}
+
+const doublesSize = function(size): numeber {
+  return size * 2;
+};
+
+const divider = {
+  curriesSize(size): number {
+    return size;
+  },
+  dividesSize: function(size): number {
+    return size / 2;
+  },
+};
+
+class Logger {
+  log(text): boolean {
+    console.log('>', text);
+    return true;
+  }
+}
+```
+
+Examples of **correct** code with `{ "parameter": true }`:
+
+```ts
+function logsSize(size: number): void {
+  console.log(size);
+}
+
+const doublesSize = function(size: number): numeber {
+  return size * 2;
+};
+
+const divider = {
+  curriesSize(size: number): number {
+    return size;
+  },
+  dividesSize: function(size: number): number {
+    return size / 2;
+  },
+};
+
+class Logger {
+  log(text: boolean): boolean {
+    console.log('>', text);
+    return true;
+  }
+}
+```
+
+### propertyDeclaration
+
+Whether to enforce type annotations for properties of interfaces and types.
+
+Examples of **incorrect** code with `{ "propertyDeclaration": true }`:
+
+```ts
+type Members = {
+  member;
+  [i: string]: ;
```

Comment:
this is not valid typescript:

```ts
interface OtherMembers {
  otherMember;
  [i: number]: ;
//             ^ Type expected.
}
```
[repl](https://www.typescriptlang.org/play/#src=interface%20OtherMembers%20%7B%0A%20%20otherMember%3B%0A%20%20%5Bi%3A%20number%5D%3A%20%3B%0A%7D)